File: src/dataset/Java/armstrong-numbers_exercism-java-blogscot/openAI/ChatGPT4_armstrong-numbers_v2.java
class ArmstrongNumbers {

  private static int[] precomputePowers(int base, int maxPower) {
    int[] powers = new int[maxPower + 1];
    powers[0] = 1;
    for (int i = 1; i <= maxPower; i++) {
      powers[i] = powers[i - 1] * base;
    }
    return powers;
  }

  boolean isArmstrongNumber(final int numberToCheck) {
    int num = numberToCheck;
    int numLength = (int) Math.log10(num) + 1;
    int[] powers = precomputePowers(9, numLength); // Precompute powers for digits 0-9
    int total = 0;

    while (num > 0) {
      int remainder = num % 10;
      total += powers[remainder];
      if (total > numberToCheck) return false; // Early exit if total exceeds number
      num /= 10;
    }
    return total == numberToCheck;
  }
}

File: src/dataset/c/armstrong-numbers_exercism-Manoloon/openAI/ChatGPT4_armstrong_numbers_v2.c
#include "armstrong_numbers.h"
#include <stdint.h>
#include <stdbool.h>

bool is_armstrong_number(int candidate)
{
    if (candidate < 0) return false;
    if (candidate < 10) return true;

    int result = 0;
    int num = candidate;
    int digits = 0;
    int powers[10] = {0};

    // Calculate the number of digits
    for (int temp = candidate; temp > 0; temp /= 10) {
        digits++;
    }

    // Precompute powers for digits 0-9
    for (int i = 0; i < 10; i++) {
        int power = 1;
        for (int j = 0; j < digits; j++) {
            power *= i;
        }
        powers[i] = power;
    }

    // Calculate Armstrong sum
    while (num > 0) {
        int r = num % 10;
        result += powers[r];
        num /= 10;
        if (result > candidate) return false; // Early exit if sum exceeds candidate
    }

    return result == candidate;
}